src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +151,150 @@ internal class DiagnosticComputer
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Step 1:
+                //  - Normal priority task must wait for all the executing high priority tasks to complete
+                //    before beginning execution.
+                await WaitForHighPriorityTasksAsync(cancellationToken).ConfigureAwait(false);
+
+                // Step 2:
+                //  - Create the core 'computeTask' for computing diagnostics.
+                var computeTask = CreateComputeTask(cancellationToken);
+
+                try
+                {
+                    // Step 3:
+                    //  - Execute the core 'computeTask' for diagnostic computation.
+                    return await computeTask.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Step 4:
+                    // Attempt to re-execute this cancelled normal priority task by running the loop again,
+                    // unless cancellation was fired on the cancellationToken passed into us.
+                    cancellationToken.ThrowIfCancellationRequested();
+                    continue;
+                }
+            }
+
+            Task<SerializableDiagnosticAnalysisResults> CreateComputeTask(CancellationToken cancellationToken)
+            {
+                return Task.Run(async () =>
+                {
+                    // Create a linked cancellation source to allow high priority tasks to cancel normal priority tasks.
+                    using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(GetCancellationTokenForNormalPriorityTasks(), cancellationToken);
+                    return await GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, linkedCancellationTokenSource.Token).ConfigureAwait(false);
+                }, cancellationToken);
+            }
+
+            static async Task WaitForHighPriorityTasksAsync(CancellationToken cancellationToken)
+            {
+                // We loop continuously until we have an empty high priority task queue.
+                while (true)
+                {
+                    ImmutableHashSet<Task> highPriorityTasksToAwait;
+                    lock (s_gate)
+                    {
+                        highPriorityTasksToAwait = s_highPriorityComputeTasks;
+                    }
+
+                    if (highPriorityTasksToAwait.IsEmpty)
+                    {
+                        return;
+                    }
+
+                    foreach (var task in highPriorityTasksToAwait)
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        // Wait for the high priority task, ignoring all exceptions from it.
+                        await task.NoThrowAwaitable(false);
```

Comment:
```suggestion
                        await task.WithCancellation(cancellationToken).NoThrowAwaitable(false);
```

Reply:
this ensures that if the normal pri task it canceled, it bails out immediately, even if the high pri task is taking a long time.

Alternatively, you can write `await Task.WhenAll(highPriorityTasksToAwait).WithCancellation(cancellationToken).NoThrowAwaitable(false);`